### PR TITLE
improve ca cmp

### PR DESCRIPTION
**Pull request by fspmarshall**
The `CompareAndSwapCertAuthorities` method currently relies upon deterministic serialization to function correctly.  Until #10070  it was possible that serialization wouldn't actually be deterministic.  After that PR was merged, @espadolini discovered another potential failure case.  If a new field is added to the ca resource and the auth server must be downgraded for some reason, the comparison fails due to the new field being omitted.

This PR changes `CompareAndSwapCertAuthorities` to instead load the value that actually exists at the target key, perform comparison using deserialized values (avoiding downgrade issues), and pass the original `backend.Item` back to `Backend.CompareAndSwap` (further mitigating the risk of non-deterministic serialization).

### CODE REVIEW

File: lib/services/local/trust.go [hunk 1]
```diff
@@ -93,42 +91,42 @@ func (s *CA) UpsertCertAuthority(ca types.CertAuthority) error {
 	return nil
 }
 
-const compareAndSwapFixExample = "tctl get %s/%s/%s --with-secrets > ca.yaml && tctl create -f ca.yaml && rm ca.yaml"
-
 // CompareAndSwapCertAuthority updates the cert authority value
 // if the existing value matches existing parameter, returns nil if succeeds,
```
Comment (espadolini) [inline lib/services/local/trust.go:95]:
```suggestion
// if the existing value matches expected parameter, returns nil if succeeds,
```

File: lib/services/local/trust.go [hunk 2]
```diff
@@ -93,42 +91,42 @@ func (s *CA) UpsertCertAuthority(ca types.CertAuthority) error {
...
-		Key:     backend.Key(authoritiesPrefix, string(new.GetType()), new.GetName()),
-		Value:   newValue,
-		Expires: new.Expiry(),
+	actual, err := services.UnmarshalCertAuthority(actualItem.Value)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !services.CertAuthoritiesEquivalent(actual, expected) {
+		return trace.CompareFailed("cluster %v settings have been updated, try again", new.GetName())
```
Comment (espadolini) [inline lib/services/local/trust.go:114]: I realize that it's the exact same message as before, but this is only accurate if the rotation was indeed the only thing different between actual and expected. Would it be too much of a layer violation if we checked for that here?

**Review by espadolini — APPROVED** [review]

File: lib/services/local/trust.go (hunk 2 above)
Comment (fspmarshall) [reply lib/services/local/trust.go:114]: Not sure I follow.  Are you suggesting we return a different error message based on which elements of the CAs differ?

Comment (fspmarshall) [reply lib/services/local/trust.go:114]: E.g. check rotation state and return something like `"rotation state concurrently updated"`?

Comment (espadolini) [reply lib/services/local/trust.go:114]: Yes, and maybe leave the "try again" as that message is displayed verbatim by `tctl` on a manual auth - whereas if something bad has happened and we are never actually going to recover by waiting, there's no point in telling the user to try again.

Comment (espadolini) [reply lib/services/local/trust.go:114]: Admittedly the fix could be to actually catch the error on one of the layers between this and the user instead

**Review by r0mant — APPROVED** [review]

**Comment by russjones** [comment]
@fspmarshall Looks like this is approved, can we merge?

**Pull request merged**